File: Objective-C/LRI/Drink MVP/Presenter/DrinkPresenter.m
//
//  DrinkPresenter.m
//  LRI
//
//  Created by Development on 05.07.2022.
//

#import <Foundation/Foundation.h>
#import "DrinkPresenter.h"
#import "DrinkEndpoint.h"
#import "Drinkmodel.h"

@interface DrinkPresenter ()
@property (nonatomic, weak) id <DrinkViewProtocol> delegate;
@property (nonatomic, strong) id <NetworkDataFetchableProtocol> networkAPI;
@property (nonatomic, strong) DrinkModel *model;
@end

@implementation DrinkPresenter

- (instancetype)initWithViewDelegate:(id <DrinkViewProtocol>) delegate withNetworkAPI:(id <NetworkDataFetchableProtocol>) networkAPI {
    self = [super init];
    if (self) {
        self.delegate = delegate;
        self.networkAPI = networkAPI;
        self.model = [[DrinkModel alloc] init];
    }
    return self;
}

- (nonnull NSMutableArray<DrinkItem *> *)getItems {
    return [self.model getItems];
}

- (NSInteger)getCount {
    return [self.model getCount];
}

- (void)getDrinks {
    DrinkEndpoint *endpoint = [[DrinkEndpoint alloc] init];
    [self.networkAPI performRequestWithEndpoint:endpoint withCompletion:^(NSMutableArray<DrinkItem *> *items, NSError *error) {
        __weak typeof(self) weakSelf = self;
        if (error == nil) {
            [weakSelf.model fillDataWith:items];
            [weakSelf.delegate presentDrinks];
        } else {
            NSString *title = @"Something went wrong";
            NSString *message = [error.userInfo valueForKey:@"NSLocalizedDescription"];
            [weakSelf.delegate presentAlertWith:title withMessage:message];
        }
    }];
}

@end

File: SwiftUIKit/LRI/Drink MVP/Presenter/DrinkPresenter.swift
//
//  DrinkPresenter.swift
//  LRI
//
//  Created by Development on 04.07.2022.
//

import Foundation

protocol DrinkViewProtocol: AnyObject {
    func presentDrinks()
    func presentAlert(with title: String?, message: String)
}

protocol DrinkViewPresenterProtocol: AnyObject {
    var items: [DrinkItem] { get }
    var count: Int { get }
    
    func getDrinks()
}

class DrinkPresenter {
    private weak var delegate: DrinkViewProtocol?
    private let networkAPI: NetworkDataFetchableProtocol
    private let model: DrinkModel
    
    required init(view delegate: DrinkViewProtocol,
                  networkAPI: NetworkDataFetchableProtocol) {
        self.delegate = delegate
        self.networkAPI = networkAPI
        self.model = DrinkModel()
    }
}

extension DrinkPresenter: DrinkViewPresenterProtocol {
    /// Drink items for representing on table view controller
    var items: [DrinkItem] {
        self.model.items
    }
    /// Drink items count
    var count: Int {
        self.model.count
    }
    /// Fetch drink items
    func getDrinks() {
        self.networkAPI.performRequest(endpoint: DrinkEndpoint.getDrinks) { [weak self] (result: Result<DrinkResponse, Error>) in
            switch result {
            case .success(let response):
                let drinks = response.drinks
                self?.model.fillData(with: drinks)
                self?.delegate?.presentDrinks()
            case .failure(let error):
                let title = "Something went wrong"
                self?.delegate?.presentAlert(with: title, message: error.localizedDescription)
            }
        }
    }
}
